File: app/src/main/java/com/savvasdalkitsis/gameframe/draw/model/ScratchTool.java
package com.savvasdalkitsis.gameframe.draw.model;

import com.savvasdalkitsis.gameframe.grid.model.Grid;

abstract class ScratchTool extends AbstractDrawingTool {

    @Override
    public void drawOn(Layer layer, int startColumn, int startRow, int column, int row, int color) {
        Grid scratch = layer.getColorGrid().initializeScratch();
        drawOnScratch(scratch, startColumn, startRow, column, row, color);
    }

    abstract void drawOnScratch(Grid scratch, int startColumn, int startRow, int column, int row, int color);

    @Override
    public void finishStroke(Layer layer) {
        layer.getColorGrid().rasterScratch();
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/draw/model/ScratchTool.kt
package com.savvasdalkitsis.gameframe.feature.draw.model

import com.savvasdalkitsis.gameframe.feature.grid.model.Grid

abstract class ScratchTool : AbstractDrawingTool() {

    override fun drawOn(layer: Layer, startColumn: Int, startRow: Int, column: Int, row: Int, color: Int) {
        val scratch = layer.colorGrid.initializeScratch()
        drawOnScratch(scratch, startColumn, startRow, column, row, color)
    }

    internal abstract fun drawOnScratch(scratch: Grid, startColumn: Int, startRow: Int, column: Int, row: Int, color: Int)

    override fun finishStroke(selectedLayer: Layer) {
        selectedLayer.colorGrid.rasterScratch()
    }
}
